File: backend/models/backtest.py
"""Backtest request + response schemas.

Small, explicit pydantic v2 models so FastAPI does the shape validation and
the Next.js frontend gets a stable contract. Matches the shape returned by
``services.backtest_service.run_backtest``.
"""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class BacktestRequest(BaseModel):
    """POST /api/backtest body."""

    entry_z: float = Field(2.0, description="Absolute Z-score that triggers entry.")
    exit_z: float = Field(0.2, description="Absolute Z-score that triggers flat.")
    lookback_days: int = Field(
        365,
        gt=30,
        le=3650,
        description="History window fed into the backtest.",
    )
    slippage_per_bbl: float = Field(
        0.0, ge=0.0, description="Per-barrel slippage applied on each fill."
    )
    commission_per_trade: float = Field(
        0.0, ge=0.0, description="Flat USD commission charged per leg fill."
    )


class EquityPoint(BaseModel):
    """Single point on the cumulative-PnL equity curve."""

    Date: Optional[str] = None
    cum_pnl_usd: Optional[float] = None


class BacktestResponse(BaseModel):
    """POST /api/backtest response."""

    sharpe: Optional[float] = None
    sortino: Optional[float] = None
    calmar: Optional[float] = None
    var_95: Optional[float] = None
    es_95: Optional[float] = None
    max_drawdown: Optional[float] = None
    hit_rate: Optional[float] = None
    total_pnl_usd: Optional[float] = None
    n_trades: int = 0
    avg_days_held: Optional[float] = None
    avg_pnl_per_bbl: Optional[float] = None
    rolling_12m_sharpe: Optional[float] = None
    equity_curve: list[dict] = Field(default_factory=list)
    trades: list[dict] = Field(default_factory=list)
    params: dict = Field(default_factory=dict)


__all__ = ["BacktestRequest", "BacktestResponse", "EquityPoint"]

File: frontend/__tests__/fixtures/backtest.ts
import type { BacktestLiveResponse } from "@/types/api";

/**
 * Deterministic fixture mimicking POST /api/backtest — a 200-bar
 * cumulative-PnL curve with plausible drawdowns.
 */
export function makeBacktestFixture(points = 200): BacktestLiveResponse {
  let cum = 0;
  const equity_curve = Array.from({ length: points }, (_, i) => {
    const d = new Date("2025-01-01T00:00:00Z");
    d.setUTCDate(d.getUTCDate() + i);
    const bar = Math.sin(i / 9) * 800 + 120;
    cum += bar;
    return {
      Date: d.toISOString().slice(0, 10),
      cum_pnl_usd: cum,
    };
  });
  return {
    sharpe: 1.42,
    sortino: 1.91,
    calmar: 0.87,
    var_95: -1800,
    es_95: -2400,
    max_drawdown: -4200,
    hit_rate: 0.58,
    total_pnl_usd: cum,
    n_trades: 24,
    avg_days_held: 8.3,
    avg_pnl_per_bbl: 0.41,
    rolling_12m_sharpe: 1.38,
    equity_curve,
    trades: [],
    params: { entry_z: 2.0, exit_z: 0.2, lookback_days: 365 },
  };
}
